src/clj/chronograph/handlers/google_auth.clj

```diff
@@ -0,0 +1,85 @@
+(ns chronograph.handlers.google-auth
+  (:require [ring.util.codec :as codec]
+            [ring.util.response :as response]
+            [chronograph.config :as config]
+            [chronograph.auth :as auth]
+            [chronograph.domain.user :as user]
+            [org.httpkit.client :as http]
+            [mount.core :refer [defstate]]
+            [cheshire.core :as json]
+            [taoensso.timbre :as log])
+  (:import (com.google.api.client.googleapis.auth.oauth2 GoogleIdTokenVerifier$Builder GoogleIdTokenVerifier)
+           (com.google.api.client.googleapis.javanet GoogleNetHttpTransport)
+           (com.google.api.client.json.jackson2 JacksonFactory)))
+
+(defstate google-id-token-verifier
+  :start (-> (GoogleIdTokenVerifier$Builder. (GoogleNetHttpTransport/newTrustedTransport)
+                                             (JacksonFactory.))
+             (.setAudience [(get-in config/config [:oauth :google :client-id])])
+             (.build))
+  :stop nil)
+
+(defn verify-id-token
+  [^String token]
+  (some->> token
+           (.verify ^GoogleIdTokenVerifier google-id-token-verifier)
+           (.getPayload)
+           (into {})))
+
+(defn- redirect-uri [{:keys [redirect-uri
+                             client-id
+                             response-type
+                             scope
+                             login-endpoint] :as oauth-config}]
+  (format "%s?redirect_uri=%s&client_id=%s&scope=%s&response_type=%s"
+          login-endpoint
+          (codec/url-encode redirect-uri)
+          (codec/url-encode client-id)
+          (codec/url-encode scope)
+          (codec/url-encode response-type)))
+
+(defn login-handler [request]
+  (response/redirect (-> config/config
+                         (get-in [:oauth :google])
+                         redirect-uri)))
+
+(defn- redirect-with-error [error]
+  (response/redirect (format "/?%s=%s"
+                             (codec/url-encode "auth-error")
+                             (codec/url-encode error))))
+
+(defn- fetch-id-token [auth-code]
+  (let [{:keys [token-endpoint
+                client-id
+                client-secret
+                redirect-uri]} (get-in config/config [:oauth :google])
+        {:keys [status body]} @(http/post token-endpoint
+                                          {:form-params {"client_id"     client-id
+                                                         "client_secret" client-secret
+                                                         "code"          auth-code
+                                                         "grant_type"    "authorization_code"
+                                                         "redirect_uri"  redirect-uri}})]
+    (if-not (= 200 status)
+      (throw (ex-info "Received an unexpected status code from Google's token endpoint"
+                      {:status status
```

Comment:
Will this status be nil when the connection is unexpectedly terminated?

Reply:
Yes, because http-kit doesn't actually throw exceptions by itself. Addressed in the latest patchset.